Sources/idt/idt.cc

```diff
@@ -238,21 +271,146 @@ class visitor : public clang::RecursiveASTVisitor<visitor> {
+        return is_symbol_exported(RD);
+
     return false;
   }
 
-  // Determine if a tagged type needs exporting at the record level.
-  // Returns true if the record has been, or was already, annotated. Returns
-  // false otherwise.
-  bool export_record_if_needed(clang::CXXRecordDecl *RD) {
+  // Determine if a function needs exporting and add the export annotation as
+  // required.
+  void export_function_if_needed(const clang::FunctionDecl *FD) {
+    // Check if the symbol is already exported.
+    if (is_symbol_exported(FD))
+      return;
+
+    // Ignore declarations from the system.
+    if (is_in_system_header(FD))
+      return;
+
+    // Skip declarations not in header files.
+    if (!is_in_header(FD))
+      return;
+
+    // We are only interested in non-dependent types.
+    if (FD->isDependentContext())
+      return;
+
+    // If the function has a body, it can be materialized by the user.
+    if (FD->hasBody())
+      return;
+
+    // Skip methods in template declarations.
+    if (FD->getTemplateInstantiationPattern() != nullptr)
+      return;
+
+    // Ignore friend declarations.
+    if (FD->getFriendObjectKind() != clang::Decl::FOK_None)
+      return;
+
+    // Ignore deleted and defaulted functions (e.g. operators).
+    if (FD->isDeleted() || FD->isDefaulted())
+      return;
+
+    // Skip template class template argument deductions.
+    if (llvm::isa<clang::CXXDeductionGuideDecl>(FD))
+      return;
+
+    // Pure virtual methods cannot be exported.
+    if (const auto *MD = llvm::dyn_cast<clang::CXXMethodDecl>(FD))
+      if (MD->isPureVirtual())
+        return;
+
+    // Ignore known forward declarations (builtins)
+    if (contains(kIgnoredBuiltins, FD->getNameAsString()))
+      return;
+
+    // TODO(compnerd) replace with std::set::contains in C++20
+    if (contains(get_ignored_symbols(), FD->getNameAsString()))
+      return;
+
+    clang::SourceLocation SLoc =
+        FD->getTemplatedKind() == clang::FunctionDecl::TK_NonTemplate
+            ? FD->getBeginLoc()
+            : FD->getInnerLocStart();
+    unexported_public_interface(get_location(FD), FD)
+        << FD << clang::FixItHint::CreateInsertion(SLoc, export_macro + " ");
+  }
+
+  // Determine if a variable needs exporting and add the export annotation as
+  // required. This only applies to extern globals and static member fields.
+  void export_variable_if_needed(const clang::VarDecl *VD) {
+    // Check if the symbol is already exported.
+    if (is_symbol_exported(VD))
+      return;
+
+    // Ignore declarations from the system.
+    if (is_in_system_header(VD))
+      return;
+
+    // Skip local variables. We are only interested in static fields.
+    if (VD->getParentFunctionOrMethod())
+      return;
+
+    // Skip static fields that have initializers.
+    if (VD->hasInit())
+      return;
+
+    // Skip all variable declarations not in header files.
+    if (!is_in_header(VD))
```

Comment:
Should we hoist this to after the system header check? That mirrors what we do in the function check.